### PR TITLE
make setup_oracle_clientfail gracefully with useful log message

### LINKED ISSUE #76: Make setup_oracle_client fail gracefully if zip file is unavailable
### Summary

As an etlhelper user, I want a helpful message if `setup_oracle_client` cannot unpack the zip file, so that I can understand what it wrong without digging through stack traces.

### Description

The `setup_oracle_client` function called by the `setup_oracle_client` script crashes with a stack trace if there is a problem downloading or opening the zip file.  This is unhelpful for users as the cause of failure may not be obvious.  It would be better to fail gracefully.

This could be achieved by putting the call to `_install_instantclient` within a `try... except` block that catches the exception and calls `sys.exit(1)`.  If helpful failure messages are raised by the downloading/unzipping functions, their text could be logged after they are caught, just before exiting.

https://github.com/BritishGeologicalSurvey/etlhelper/blob/84661a1bf78f17f42327f43ce77c0ea031007374/etlhelper/setup_oracle_client.py#L57

#### Possible implementation

```python
try:
    _install_instantclient(zip_location, install_dir, ld_library_prepend_script)
except (SomeError, AnotherError) as e:
    logger.error(e.args[0])
    sys.exit(1)
```

### Acceptance criteria

+ [x] Failure because the URL cannot be downloaded causes script to exit with helpful log message
+ [x] Failure because file path doesn't exist causes script to exit with helpful log message
+ [x] Failure because file is not a zip file causes script to exit with helpful log message
+ [x] `_download_zipfile` docstring is updated
+ [x] All failure modes are covered by integration tests

### PR DESCRIPTION
**Pull request by kerberpolis**
resolve #76 

urllib.request captures URLError and raises it for higher try/except block in setup_oracle_client

### CODE REVIEW

File: etlhelper/setup_oracle_client.py [hunk 1]
```diff
@@ -169,14 +178,16 @@ def _download_zipfile(zip_download_source):
         zip_download_source
     returns:
         path to downloaded zipfile
-
-    Fails gracefully if download fails. Sys exit -1 with helpful msg
```
Comment (dvalters) [inline etlhelper/setup_oracle_client.py:173]: I think "fails gracefully...etc" can be left in, just needs the "Sys exit" bit removed

**Review by dvalters — CHANGES_REQUESTED** [review]
Looks good, one tiny comment about the docstring. It only works if the `--reinstall` flag is given, otherwise it says the oracle client is already installed, without checking for a bad zipfile or path if one is given. (Is that correct behaviour @volcan01010?)

**Comment by volcan01010** [comment]
Yes, if you already have an installation of the oracle client, the new changes will only take effect if you call with reinstall.  Otherwise you could delete some files to break your existing installation and that will make it attempt installation again, too.

I gave it a quick spin just now.  The error message for a bad url is wrong when the server exists and returns a 404 page.  When I tried it, it seemed to download the page with the 404 message as if it was the zip file.  Then I get an "is not a valid zip file" error instead.  Can you check what is happening with the errors - does a 404 definitely raise a URLError?

```
$ git checkout setup_script_fail_gracefully 
$ pip install -e .

$ setup_oracle_client -z /tmp/this-does-not-exist
setup_oracle_client: Zip path /tmp/this-does-not-exist does not exist

$ setup_oracle_client -z /tmp
setup_oracle_client: Zip path /tmp is not a valid zip file

$ setup_oracle_client -z http://bad.url
setup_oracle_client: <urlopen error [Errno -2] Name or service not known>

$ setup_oracle_client -z http://www.bgs.ac.uk/bad-url
setup_oracle_client: Zip path /tmp/bad-url is not a valid zip file
```

"zip_location" instead of "Zip path" in the message would make it consistent with `setup_oracle_client --help`.

**Comment by kerberpolis** [comment]
> Yes, if you already have an installation of the oracle client, the new changes will only take effect if you call with reinstall. Otherwise you could delete some files to break your existing installation and that will make it attempt installation again, too.
> 
> I gave it a quick spin just now. The error message for a bad url is wrong when the server exists and returns a 404 page. When I tried it, it seemed to download the page with the 404 message as if it was the zip file. Then I get an "is not a valid zip file" error instead. Can you check what is happening with the errors - does a 404 definitely raise a URLError?
> 
> ```
> $ git checkout setup_script_fail_gracefully 
> $ pip install -e .
> 
> $ setup_oracle_client -z /tmp/this-does-not-exist
> setup_oracle_client: Zip path /tmp/this-does-not-exist does not exist
> 
> $ setup_oracle_client -z /tmp
> setup_oracle_client: Zip path /tmp is not a valid zip file
> 
> $ setup_oracle_client -z http://bad.url
> setup_oracle_client: <urlopen error [Errno -2] Name or service not known>
> 
> $ setup_oracle_client -z http://www.bgs.ac.uk/bad-url
> setup_oracle_client: Zip path /tmp/bad-url is not a valid zip file
> ```
> 
> "zip_location" instead of "Zip path" in the message would make it consistent with `setup_oracle_client --help`.

I also saw the same issue on Tuesday when I recreated the issue but as of today the URLError is caught and returns the 404 message. I think they changed the response type.

There is the same issue with the response message but for other endpoints. I think some endpoints get redirected instead of returning 404. I've checked this with other url endpoints and get redirected to different news articles or staff profiles on the bgs website. For example I tried http://www.bgs.ac.uk/sutt which returns my staff profile under /people/sutton-daniel. This downloads the html to /tmp/sutt and fails the zip file checks which is what was happening to you with /bad-url.

I wonder if we could use [Requests](https://2.python-requests.org/en/master/user/quickstart/#redirection-and-history) over urllib as they can easily handle redirections? urllib can also do that but would have to [extend urllib.request.HTTPRedirectHandler](https://stackoverflow.com/questions/52086805/how-to-disable-automatic-redirects-in-python3-urllib-request).

**Comment by volcan01010** [comment]
I'd tried to use `urllib` to limit the number of dependencies that `etlhelper` has, but maybe that doesn't matter too much.

I think a redirect is OK.  In real life a redirect may take you to a zip file e.g. the latest version.  The redirect for the `dsutt` example is the correct behaviour, and if there was nothing at the redirected page you should get the 404.

Does this raise the 404 / URLError error?

+ https://http.cat/40404040

If so, then I think that we are good.

**Comment by kerberpolis** [comment]
Yep that raises the URLError and returns a 404 message. I'll finish up the tests to make sure it returns the correct logging messages.

File: etlhelper/setup_oracle_client.py (hunk 1 above)
Comment (volcan01010) [reply etlhelper/setup_oracle_client.py:173]: I think this needs approval from @dvalters to resolve.  I think it's OK to remove the whole line.

**Comment by volcan01010** [comment]
I've added two minor tweaks to this.  I think it is ready to merge now.

**Pull request merged**